spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRowLevelOperationsWithLineage.java

```diff
@@ -0,0 +1,454 @@
+  static final Schema SCHEMA =
+      new Schema(
+          ImmutableList.of(
+              Types.NestedField.required(1, "data", Types.IntegerType.get()),
+              MetadataColumns.ROW_ID,
+              MetadataColumns.LAST_UPDATED_SEQUENCE_NUMBER));
+
+  static final List<Record> INITIAL_RECORDS =
+      ImmutableList.of(
+          createRecord(0, 0L, 1L),
+          createRecord(1, 1L, 1L),
+          createRecord(2, 2L, 1L),
+          createRecord(3, 3L, 1L),
+          createRecord(4, 4L, 1L));
+
+  @BeforeAll
+  public static void setupSparkConf() {
+    spark.conf().set("spark.sql.shuffle.partitions", "4");
+  }
+
+  @BeforeEach
+  public void beforeEach() {
+    assumeThat(formatVersion).isGreaterThanOrEqualTo(3);
+    // ToDo: Remove these as row lineage inheritance gets implemented in the other readers
+    assumeThat(fileFormat).isEqualTo(FileFormat.PARQUET);
+    assumeThat(vectorized).isFalse();
+  }
+
+  @AfterEach
+  public void removeTables() {
+    sql("DROP TABLE IF EXISTS %s", tableName);
+    sql("DROP TABLE IF EXISTS source");
+  }
+
+  @TestTemplate
+  public void testMergeIntoWithBothMatchedAndNonMatched()
+      throws NoSuchTableException, ParseException, IOException {
+    assumeThat(formatVersion).isGreaterThanOrEqualTo(3);
+
+    createAndInitTable("data INT", null);
+    createBranchIfNeeded();
+    Table table = loadIcebergTable(spark, tableName);
+    appendRecords(table, partitionRecords(INITIAL_RECORDS, table.spec(), UNPARTITIONED_GENERATOR));
+    createOrReplaceView("source", "data INT", "{ \"data\": 1}\n" + "{ \"data\": 5678}\n");
+    sql(
+        "MERGE INTO %s AS t USING source AS s "
+            + "ON t.data == s.data "
+            + "WHEN MATCHED THEN "
+            + "  UPDATE SET t.data = 1234 "
+            + "WHEN NOT MATCHED THEN "
+            + "  INSERT *",
+        commitTarget());
+
+    long updateSequenceNumber = latestSnapshot(table).sequenceNumber();
+    List<Object[]> rowsWhereIdShouldBePreserved =
+        sql(
+            "SELECT data, _row_id, _last_updated_sequence_number FROM %s WHERE data != 5678 ORDER BY _row_id",
+            selectTarget());
+    assertEquals(
+        "Rows which are carried over or updated should have expected lineage",
+        ImmutableList.of(
+            row(0, 0L, 1L),
+            row(1234, 1L, updateSequenceNumber),
+            row(2, 2L, 1L),
+            row(3, 3L, 1L),
+            row(4, 4L, 1L)),
+        rowsWhereIdShouldBePreserved);
+
+    // New row with data 5678 should have any row ID higher than the max in the previous commit
+    // version (4), and have the latest seq number
+    long previousHighestRowId = 4;
+    List<Object[]> newRows =
+        sql(
+            "SELECT data, _row_id, _last_updated_sequence_number FROM %s WHERE data = 5678 ORDER BY _row_id",
+            selectTarget());
+    assertAddedRowLineage(newRows, 5678, previousHighestRowId, updateSequenceNumber);
```

Comment:
There may be a simpler way to express this without the filtering in the query...we can get all the records sorted by row ID. Since we know the new record must be at the end we can sublist every element except the last, and assert that. Then take the new record and assert the same stuff as 135 and below.